universe/ShipDesign.h

```diff
@@ -520,15 +551,26 @@ class FO_COMMON_API ShipDesign {
     void                            Rename(const std::string& name) { m_name = name; }  ///< renames this design to \a name
     //@}
 
-    ///< returns true if the \a hull and parts vectors passed make a valid ShipDesign
-    static bool                     ValidDesign(const std::string& hull, const std::vector<std::string>& parts);
-
-    /** returns true if the \a design passed is a valid ShipDesign in terms of
-      * (its hull and parts.  does not check any other member variables */
-    static bool                     ValidDesign(const ShipDesign& design)
-    { return ValidDesign(design.m_hull, design.m_parts); }
-
+    /** Return true if \p hull and \p parts would make a valid design. */
+    static bool ValidDesign(const std::string& hull, const std::vector<std::string>& parts);
 private:
+    /** Return a valid hull and parts pair iff the \p hull and \p parts vectors
+        would not make a valid ShipDesign.
+        Also pad parts with "" if it is shorter than the \p hull number of slots.
+        Otherwise return none. If \p produce_log is true then produce log messages. */
+    static boost::optional<std::pair<std::string, std::vector<std::string>>>
+        MaybeInvalidDesign(const std::string& hull, std::vector<std::string>& parts, bool produce_log);
```

Comment:
Yes.  That is what the "Also pad parts with \"\"" portion of the comment refers to.

Since we decided that a short parts vector is not a warning case, and if the return value is not none it indicates an error `MaybeInvalidDesign` needs a way to adjust parts (make it as long as the hull's number of slots).